### ISSUE TITLE
About Some interesting predict result between two different datasets

### ISSUE THREAD

**Issue by ysssgdhr** [MIC-DKFZ/nnUNet#277]
Hello Fabian!
I've been working on transfer learning(unsupervised domain transfer) about kidney CT images. 
We collected more than 200 abdominal CT images from Chinese hospitals (let's call them OURS dataset). Different from KITS, we separate the left kidney and the right kidney and label them independently. The resolution of each CT image is 200*150*150 (much smaller than KITS). Moreover, **our CT images are isotropic, and the voxel spacing in Z axis is less than 1.**

As we know, the z-axis voxel spacing of CT images in KiTS dataset varies from 0.5 to 3 or even 5, which leads to the generalization and robustness of 3D fullres data. 
![image](https://user-images.githubusercontent.com/24486813/88475922-d7103980-cf66-11ea-9a0e-c616b115eb20.png)

The models trained by KiTS can achieve 91.6% (kidney) and 86.6% (tumor) in OURS dataset which is very good.
![image](https://user-images.githubusercontent.com/24486813/88476025-7af9e500-cf67-11ea-9573-26fdd74c67a2.png)

This surprised us, because we used contrast media longer, which led to clearer cortex. And our equipment is also different. Anyway, the representation of our data set should be quite different from KiTS data set, but it still achieved quite good results (which is also attributed to the powerful performance of nnUnet).

Then we tried to do the opposite, using the models trained by OURS data set to predict KiTS data set. The result is very surprising, dice is 3% and 0.8% respectively. 
![image](https://user-images.githubusercontent.com/24486813/88476001-60c00700-cf67-11ea-985c-37ec508ee25f.png)

The predicted recall rate is OK (indicating that some tissues have been detected), but the accuracy rate is extremely low. The red circle in the picture is part of the kidney tissue that is correctly predicted.
![image](https://user-images.githubusercontent.com/24486813/88476039-a11f8500-cf67-11ea-8147-16c66342feba.png)

I have two questions:
1. Is it possible that such a poor result is due to the three-axis sequence in prediction? I noticed that the plans.pkl of the KiTS model is as follows:
![image](https://user-images.githubusercontent.com/24486813/88476072-07a4a300-cf68-11ea-9f7f-930c8672453b.png)

And OURS is like this:
![image](https://user-images.githubusercontent.com/24486813/88476079-11c6a180-cf68-11ea-80a8-9aa30a54dc5c.png)

2. Do you think it is possible? The results of the two migration directions differ by 90%. I mean, KiTS data set is obviously more comprehensive, which requires higher generalization of the model. But the result of migrating from our data set to KiTS is unacceptable. (This is not a bad thing, which gives us the goal of unsupervised domain migration)

**Comment by FabianIsensee** [comment]
Hi @ysssgdhr ,
really interesting observations, thank you for sharing! I have a couple of thoughts about what may be happening.

Your comment 1) is certainly contributing a lot to a potential drop in performance. KiTS seems to have a different axis ordering than OURS (indicated by a different transpose_forward in this case (but not always)). It is critically important that the scans you are predicting have the same axis ordering than nnU-Net. What is mean by this is that when you open up the images with sitk or nibabel and export the numpy arrays, the axes need to match (for example the first coordinate should always be axial slices, so a[12] is the axial slice id 12). If they don't the performance can degrade a lot. I did run into this issue myself when I was applying KiTS to some internal CT data.

It is interesting that the transfer of KiTS -> OURS is still successful, but I think this totally *can* happen.

Even though I do not know the OURS dataset, I have some ideas what could cause the issues in transferring the model. 

First of all, please make sure the axis ordering is the same between the datasets! This can be critical, especially given the following.
Second, you should keep in mind that the transfer of a trained model between datasets can not be expected to be equally good in both directions. If one datasets is much more difficult than the other, then transferring from difficult -> easy will work very well than the other way around. You mentioned that your contrast agent was used longer and that tumors should be easier to detect. If you train on that then the model will have troubles with the more difficult KiTS dataset. Another aspect that is critical is homogeneity. If OURS is very homogeneous, then the model will not learn to cope with resampling artifacts and muddiness in the case of low resolution images (for example 5mm slices in KiTS).

So in summary, please make sure the axis ordering matches between the datasets. This will give the largest jump in performance (and may even improve upon the KiTS-> OURS result). If the transfer of OURS-> KiTS is still bad then this is because OURS may be a much easier dataset that is potentially very homogeneous, resulting in a model that does not generalize as well. 

I would be very interested in hearing how the performance changes with corrected axis ordering :-)

Best,
Fabian

**Comment by ysssgdhr** [comment]
Thank you very much for your suggestion. The axial sequence of CT images of OURS data set is (z,y,x), while KiTS is (x,y,z). 
After I am quite sure that I have changed the three axes of KiTS CT images (as shown in the figure), the predicted results are still not optimistic. Those square-shaped artifacts still exist.
origin CT:
![image](https://user-images.githubusercontent.com/24486813/88526316-616fa080-d02e-11ea-9b17-2e398cbd944f.png)
after transposed:
![image](https://user-images.githubusercontent.com/24486813/88526357-70565300-d02e-11ea-8830-e441ab8ddfa1.png)
predicted mask:
![image](https://user-images.githubusercontent.com/24486813/88526572-ae537700-d02e-11ea-9b08-15b99350833a.png)


Actually, there are two points I didn't mention above.
1. because there is enough video memory for training on TeslaV100, I modified the automatically-generated plan file by nnUnet on OURS. I changed the patch_size and batch_size from (160, 128, 112) and 2 to (160, 128, 128) and 8, and renamed the plan file  as' Ours_plan.pkl'.
![image](https://user-images.githubusercontent.com/24486813/88526830-038f8880-d02f-11ea-921b-c8fc4be41256.png)

2. The single model trained on OURS has not finished. I'm very sorry about this, because the training is slow and it hasn't finished yet.
![image](https://user-images.githubusercontent.com/24486813/88526860-0e4a1d80-d02f-11ea-9e05-ef4a5e2b5fc2.png)

I renamed the best_model file and best_plan file with approximate 200 epoch as 'model_final_checkpoint.model' and 'model_final_checkpoint.model.pkl'. Skipping the command 'nnUNet_find_best_configuration' and directly use "nnUNet_predict -i *** -o *** -tr nnUNetTrainerV2 -ctr nnUNetTrainerV2CascadeFullRes -m 3d_fullres **-p Ours** --disable_tta -t Task102_Ours" for predicting.

I would like to ask if it is possible that the poor prediction results are caused by my revised plan file and incomplete training？Because the prediction result does not improve visibly after I transposed the axis order.......

**Comment by FabianIsensee** [comment]
Hi,
sorry for keeping you waiting for so long. I think it would make sense to investigate more thoroughly what cases this issue. I still believe that either the axis ordering OR the intensity values are at fault. I also found some inconsistencies in your previous message:
- why are you specifying `-ctr nnUNetTrainerV2CascadeFullRes` when you are using `-m 3d_fullres`?
- why are you changing the patch size along with the batch size?
- the new name of the plans file is not correct and nnU-Net will not find it as is. It needs to be OURS_plans_3D.pkl, not OURS_plans.pkl.

Here is some suggestions that should allow you to improve the situation and if not at least speed up your workflows significantly:
1) the difference in Dice scores between the configurations is extremely slow. Since your results point to a very big problem, it may make sense to use a simpler model. Please run the experiments with 3d_lowres only
2) please do not tamper with the plans files. The training is much faster when running with batch size 2 (I can get this thing trained in about 2 days on a single GPU) and there is virtually no benefit in increasing the batch size at this point
3) please compare the computed values for intensity preprocessing by using the following snippet
```python
from batchgenerators.utilities.file_and_folder_operations import *
from nnunet.paths import nnUNet_raw_data
a = load_pickle(join(nnUNet_raw_data, 'nnUNet_cropped_data/Task040_KiTS/dataset_properties.pkl'))
print('mean', a['intensityproperties'][0]['mean'])
print('mean', a['intensityproperties'][0]['sd'])
print('mean', a['intensityproperties'][0]['percentile_99_5'])
print('mean', a['intensityproperties'][0]['percentile_00_5'])
```
Run this for both datasets and let me know the values

Please also provide
```python
a['all_sizes']
a['all_spacings']
```

You can also just send me the two dataset_properties files.

Best,
Fabian

**Comment by FabianIsensee** [comment]
Some more thoughts:
- the results look so bad that I don't think transpose_forward could be the problem
- please let the training finish and run inference with the finished model. Do not tamper with the checkpoint files!
- i have used nnU-Nets trained on Kits on many other CT datasets and it has always worked perfectly. So there really  is no underlying issue here. It has to have something to do with either the models you trained so far (please follow the instructions in my previous post) or with your dataset!

**Comment by FabianIsensee** [comment]
any updates?

**Comment by joeyzyz** [comment]
> It is critically important that the scans you are predicting have the same axis ordering than nnU-Net. What I mean by this is that when you open up the images with sitk or nibabel and export the numpy arrays, the axes need to match (for example the first coordinate should always be axial slices, so a[12] is the axial slice id 12). If they don't the performance can degrade a lot. I did run into this issue myself when I was applying KiTS to some internal CT data.

Hi, Fabian,

Thanks for your reply, I also have the same question similar with this issue.

You said you also had the axis ordering mismatch problem while apply KiTS model to other dataset, I am wondering:

1. how did you fix it, that is, how to revise the code in preprocessing data?
2. I got the solution by referring to [here](https://nipy.org/nibabel/image_orientation.html), It can be illustrated with one line of code:
`canonical_img = nib.as_closest_canonical(img)`，
but it turned out not working, could you share me with your solution? 

Thank you very much~
